paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala

```diff
@@ -0,0 +1,199 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.paimon.spark.commands
+
+import org.apache.paimon.data.BinaryRow
+import org.apache.paimon.index.PartitionIndex
+import org.apache.paimon.spark.SparkRow
+import org.apache.paimon.table.FileStoreTable
+import org.apache.paimon.table.sink.{BatchWriteBuilder, CommitMessageSerializer, DynamicBucketRow, InnerTableCommit, RowPartitionKeyExtractor}
+import org.apache.paimon.types.RowType
+
+import org.apache.spark.TaskContext
+import org.apache.spark.sql.{DataFrame, Row, SparkSession}
+import org.apache.spark.sql.catalyst.encoders.{ExpressionEncoder, RowEncoder}
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.execution.command.RunnableCommand
+import org.apache.spark.sql.functions._
+
+import java.util.function.IntPredicate
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+/** Used to write a [[DataFrame]] into a paimon table. */
+case class WriteIntoPaimonTable(table: FileStoreTable, overwrite: Boolean, data: DataFrame)
+  extends RunnableCommand
+  with PaimonCommand {
+
+  import WriteIntoPaimonTable._
+
+  private lazy val tableSchema = table.schema()
+
+  private lazy val rowType = table.rowType()
+
+  private lazy val writeBuilder: BatchWriteBuilder = table.newBatchWriteBuilder()
+
+  private lazy val serializer = new CommitMessageSerializer
+
+  if (overwrite) {
+    throw new UnsupportedOperationException("Overwrite is unsupported.");
+  }
+
+  override def run(sparkSession: SparkSession): Seq[Row] = {
+    import sparkSession.implicits._
+
+    val dataEncoder = RowEncoder.apply(data.schema).resolveAndBind()
+    val originFromRow = dataEncoder.createDeserializer()
+
+    val partitionCols = tableSchema.partitionKeys().asScala.map(col)
+    val primaryKeyCols = tableSchema.trimmedPrimaryKeys().asScala.map(col)
+    val partitioned = if (isDynamicBucketTable && primaryKeyCols.nonEmpty) {
+      // Make sure that the records with the same bucket values is within a task.
+      data.repartition(primaryKeyCols: _*)
+    } else {
+      data
+    }
+
+    // append _bucket_ column as placeholder
+    val withBucketCol = partitioned.withColumn(BUCKET_COL, lit(-1))
+    val numSparkPartitions = withBucketCol.rdd.getNumPartitions
+    val bucketColIdx = withBucketCol.schema.size - 1
+    val withBucketDataEncoder = RowEncoder.apply(withBucketCol.schema).resolveAndBind()
+    val toRow = withBucketDataEncoder.createSerializer()
+    val fromRow = withBucketDataEncoder.createDeserializer()
+
+    val bucketProcessor = if (isDynamicBucketTable) {
```

Comment:
Maybe here we can have a big block:
```
val bucketAssigned = if (isDynamicBucketTable) {
    if (primaryKeyCols.nonEmpty) data.repartition(primaryKeyCols: _*)
    .........
} else {
    mapPartitions(CommonBucketProcessor)....
}
```

There is no need to check isDynamicBucketTable elsewhere